simulation/src/curp_group.rs

```diff
@@ -0,0 +1,411 @@
+use std::{collections::HashMap, error::Error, path::PathBuf, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use curp::{
+    client::{Client, ReadState},
+    cmd::Command,
+    error::ProposeError,
+    members::ClusterMember,
+    server::Rpc,
+    FetchLeaderRequest, FetchLeaderResponse, LogIndex, SnapshotAllocator,
+};
+use curp_test_utils::{
+    random_id,
+    test_cmd::{TestCE, TestCommand, TestCommandResult},
+    TestRoleChange, TestRoleChangeInner,
+};
+use engine::{Engine, EngineType, Snapshot};
+use itertools::Itertools;
+use madsim::runtime::NodeHandle;
+use parking_lot::Mutex;
+use tokio::sync::mpsc;
+use tracing::debug;
+use utils::config::{ClientTimeout, CurpConfigBuilder};
+
+pub type ServerId = String;
+
+pub use curp::{
+    propose_response, protocol_client::ProtocolClient, ProposeRequest, ProposeResponse,
+};
+
+struct MemorySnapshotAllocator;
+
+#[async_trait]
+impl SnapshotAllocator for MemorySnapshotAllocator {
+    async fn allocate_new_snapshot(&self) -> Result<Snapshot, Box<dyn Error>> {
+        Ok(Snapshot::new_for_receiving(EngineType::Memory)?)
+    }
+}
+
+pub struct CurpNode {
+    pub id: ServerId,
+    pub addr: String,
+    pub handle: NodeHandle,
+    pub exe_rx: mpsc::UnboundedReceiver<(TestCommand, TestCommandResult)>,
+    pub as_rx: mpsc::UnboundedReceiver<(TestCommand, LogIndex)>,
+    pub store: Arc<Mutex<Option<Arc<Engine>>>>,
+    pub storage_path: String,
+    pub role_change_arc: Arc<TestRoleChangeInner>,
+}
+
+pub struct CurpGroup {
+    pub nodes: HashMap<ServerId, CurpNode>,
+    pub leader: Arc<Mutex<ServerId>>,
+    pub all: HashMap<ServerId, String>,
+    pub client_node: NodeHandle,
+}
+
+impl CurpGroup {
+    pub async fn new(n_nodes: usize) -> Self {
+        assert!(n_nodes >= 3);
+        let handle = madsim::runtime::Handle::current();
+
+        let all: HashMap<ServerId, String> = (0..n_nodes)
+            .map(|x| (format!("S{x}"), format!("192.168.1.{}:12345", x + 1)))
```

Comment:
Because the IP doesn't matter, both the CURP nodes and CURP client are run in the Madsim simulation environment, so it is not really an IP address. The nodes only have to know each other that's sufficient.

Reply:
got it